x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,329 @@
+// tokenGenID of the OIDCAuthenticator. It should never actually be greater than, but only equal,
+// but this is a safety check, since extra invalidation is only a performance impact, not a
+// correctness impact.
+// This must only be called with the lock held
+func (oa *OIDCAuthenticator) invalidateAccessToken() {
+	tokenGenID := oa.cfg.Connection.OIDCTokenGenID()
+	if tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		oa.cfg.Connection.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context) error {
+	oa.mu.Lock()
+	oa.invalidateAccessToken()
+	oa.mu.Unlock()
+	// it should be impossible to get a Reauth when an Auth has never occurred,
+	// so we assume cfg was properly set. There is nothing to enforce this, however,
+	// other than the current driver code flow. If cfg is nil, Auth will return an error.
+	return oa.Auth(ctx, oa.cfg)
+}
+
+// Auth authenticates the connection.
+func (oa *OIDCAuthenticator) Auth(ctx context.Context, cfg *Config) error {
+	// the Mutex must be held during the entire Auth call so that multiple racing attempts
+	// to authenticate will not result in multiple callbacks. The losers on the Mutex will
+	// retrieve the access token from the Authenticator cache.
+	oa.mu.Lock()
+	defer oa.mu.Unlock()
+	var err error
+
+	if cfg == nil {
+		return newAuthError(fmt.Sprintf("config must be set for %q authentication", MongoDBOIDC), nil)
+	}
+	oa.cfg = cfg
+
+	if oa.accessToken != "" {
+		err = ConductSaslConversation(ctx, cfg, "$external", &oidcOneStep{
+			userName:    oa.userName,
+			accessToken: oa.accessToken,
+		})
+		if err == nil {
+			return nil
+		}
+		oa.invalidateAccessToken()
+		time.Sleep(invalidateSleepTimeout)
+	}
+
+	if oa.OIDCHumanCallback != nil {
+		return oa.doAuthHuman(ctx, cfg, oa.OIDCHumanCallback)
+	}
+
+	// Handle user provided or automatic provider machine callback.
+	var machineCallback OIDCCallback
+	if oa.OIDCMachineCallback != nil {
+		machineCallback = oa.OIDCMachineCallback
+	} else {
+		machineCallback, err = oa.providerCallback()
+		if err != nil {
+			return fmt.Errorf("error getting built-in OIDC provider: %w", err)
+		}
+	}
+
+	if machineCallback != nil {
+		return oa.doAuthMachine(ctx, cfg, machineCallback)
+	}
+	return newAuthError("no OIDC callback provided", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthHuman(_ context.Context, _ *Config, _ OIDCCallback) error {
+	// TODO GODRIVER-3246: Implement OIDC human flow
+	// Println is for linter
+	fmt.Println("OIDC human flow not implemented yet", oa.idpInfo)
+	return newAuthError("OIDC human flow not implemented yet", nil)
+}
+
+func (oa *OIDCAuthenticator) doAuthMachine(ctx context.Context, cfg *Config, machineCallback OIDCCallback) error {
+	accessToken, err := oa.getAccessToken(ctx,
+		&OIDCArgs{Version: apiVersion,
```

Comment:
Optional: Format for easier reading.

```suggestion
		&OIDCArgs{
			Version: apiVersion,
```